src/ranged.cpp

```diff
@@ -161,19 +161,76 @@ bool player::handle_gun_damage( item &it, int shots_fired )
 {
     if( !it.is_gun() ) {
         debugmsg( "Tried to handle_gun_damage of a non-gun %s", it.tname() );
+        shots_fired = shots_fired + 0; // left this here in case someone needs shots_fired in the future.
         return false;
     }
-
+    int bp_jam_occurred_this_turn = 0;
+    int dirt = it.get_var( "dirt", 0 );
+    int last_fired = it.get_var( "last_fired", 0 );
+    int just_unclogged = it.get_var( "just_unclogged", 0 );
     const auto &curammo_effects = it.ammo_effects();
     const cata::optional<islot_gun> &firing = it.type->gun;
+    const std::string thistimefired = to_string_time_of_day( calendar::turn );
+    // Here we check for a chance for the weapon to suffer a misfire due to
+    // being dirty (blackpowder fouling effect only; typical dirt accumulation
+    // from smokeless powder is not currently modeled by this effect).
+    // Can damage gun, but not enough to push the weapon beyond 'shattered'.
+    // Check, but not accumulation, will happen even when firing normal rounds,
+    // so you can screw up your MP5 by firing blackpowder rounds and the
+    // effects will be in force even when firing retail ammo, though the gun
+    // won't get MORE foul.
+
+    // causes failure to cycle if weapon is fired too quickly:
+    if( last_fired == to_turn<int>( calendar::turn ) && !it.has_flag( "PUMP_ACTION" ) &&
+        !it.has_flag( "MANUAL_ACTION" ) && it.type->gun->ups_charges < 1 &&
+        ( curammo_effects.count( "MUZZLE_SMOKE" ) || curammo_effects.count( "BLACKPOWDER" ) ) ) {
+        add_msg_player_or_npc( _( "Your %s fails to cycle!" ),
+                               _( "<npcname>'s %s fails to cycle!" ),
+                               it.tname() );
+        // Don't increment until after the message
+        return false;
+    }
+    it.set_var( "last_fired", to_turn<int>( calendar::turn ) );
+    // chance to cause malfunction, which always happens when clogged:
+    int malfunctionreduction = 0;
+    if( it.has_flag( "PUMP_ACTION" ) || it.has_flag( "MANUAL_ACTION" ) ) {
+        malfunctionreduction = 30;
+    }
+    if( ( it.type->gun->ammo.count( ammotype( "flintlock" ) ) ) ||
```

Comment:
```suggestion
    if( it.type->gun->ammo.count( ammotype( "flintlock" ) ) ||
```
Same for the lines below.